core/src/price_estimation/clients/generic_client.rs

```diff
@@ -1,79 +1,91 @@
-mod api;
-
-use super::{PriceSource, TokenData};
+use super::super::{PriceSource, TokenData};
 use crate::http::HttpFactory;
 use crate::models::TokenId;
 use anyhow::{anyhow, Context, Result};
-use api::{DexagApi, DexagHttpApi};
 use futures::{
     future::{self, BoxFuture, FutureExt as _},
     lock::Mutex,
 };
 use std::collections::HashMap;
 
-struct ApiTokens {
+/// Provides a generic interface to communicate in a standardized way
+/// with specific API token implementations
+pub trait GenericToken {
+    /// Symbol describing the ERC20 token represented by the type instance
+    fn symbol(&self) -> &str;
+}
+
+#[cfg_attr(test, mockall::automock(type Token=tests::MockGenericToken;))]
+pub trait Api: Sized {
+    type Token: GenericToken + Sync + Send;
+
+    /// Creates a new HTTP interface for connecting to the API
+    fn bind(http_factory: &HttpFactory) -> Result<Self>;
+
+    fn get_token_list<'a>(&'a self) -> BoxFuture<'a, Result<Vec<Self::Token>>>;
+
+    /// Returns the price of one unit of `from` expressed in `to`.
+    /// For example `get_price("ETH", "DAI")` is ~220.
+    fn get_price<'a>(&'a self, from: &Self::Token, to: &Self::Token) -> BoxFuture<'a, Result<f64>>;
+
+    /// Returns a string representing the reference coin in the stead of OWL for this API
+    /// Could be different from "OWL", e.g., when the API does not offer prices with
+    /// respect to OWL
+    fn stable_coin_symbol() -> &'static str;
+}
+
+struct Tokens<T: Api> {
     // Maps uppercase Token::symbol to Token.
     // This is cached in the struct because we don't expect it to change often.
-    tokens: HashMap<String, api::Token>,
-    stable_coin: api::Token,
+    tokens: HashMap<String, T::Token>,
+    stable_coin: T::Token,
 }
 
-pub struct DexagClient<Api> {
-    api: Api,
+pub struct GenericClient<T: Api> {
+    api: T,
     /// Lazily retrieved the first time it is needed when `get_prices` is
     /// called. We don't want to use the network in `new`.
-    api_tokens: Mutex<Option<ApiTokens>>,
+    api_tokens: Mutex<Option<Tokens<T>>>,
     tokens: TokenData,
 }
 
-impl DexagClient<DexagHttpApi> {
-    /// Create a DexagClient using DexagHttpApi as the api implementation.
+impl<T: Api> GenericClient<T> {
+    /// Create a GenericClient using the api implementation from HttpConnecting.
     pub fn new(http_factory: &HttpFactory, tokens: TokenData) -> Result<Self> {
```

Comment:
This might be one of the best code review comments I have ever seen. I am now just a lurker/spectator tagged as a reviewer but now learning from a rust guru!

Reply:
An issue I see with this suggestion is that it's going to move some complexity down to `price_estimation.rs`.

Currently the API is already part of the client, and we only need to import:
```
use self::clients::{DexagClient, KrakenClient};
```

With the proposed changes we'd need to import all "genericized" APIs in `price_estimation.rs` and use each of them to "specificize" a generic client (which must then receive a different treatment compared to an ad hoc clients). With the current state of the PR, this complexity is hidden in the type creation in `dexag.rs` and there is no difference between ad hoc clients and clients built from `GenericClient` in `price_estimation.rs`.

Also, the only user of the clients discussed in this PR is `PriceOracle`, which is intrinsically bound to HTTP since its constructor requires an HTTP factory. So, if we need to make this PR more general it might be enough to replace every occurrence of `HttpFactory` with `impl Connector` for some custom `Connector` trait used to generalize `HttpFactory` that is specified somewhere else in the code. In general, I think if we'll need to use something that is not HTTP in the future, then the code surrounding this PR might end up looking quite different anyway. 

In conclusion, I'd not change the current code until we actually need to use an API that is not served through HTTP.